src/NuGet.Core/NuGet.PackageManagement/NuGetProject.cs

```diff
@@ -74,13 +78,36 @@ public virtual Task PostProcessAsync(INuGetProjectContext nuGetProjectContext, C
             return Task.FromResult(0);
         }
 
+        public virtual void EnsureMetadataInitialized()
+        {
+            // do nothing by default.
+        }
+
+        private void EnsureInitialized()
+        {
+            lock (_lckObj)
+            {
+                if (!_initialized)
+                {
+                    _initialized = true;
+
+                    EnsureMetadataInitialized();
+                }
+            }
+        }
+
         public T GetMetadata<T>(string key)
         {
             if (key == null)
             {
                 throw new ArgumentNullException("key");
             }
 
+            if (!_initialized)
```

Comment:
it's just a preventive way to avoid executing it twice since it's being called at 2 places.